### ISSUE TITLE
bio tools fail to compile, incomplete definition of type 'struct request'

### ISSUE THREAD

**Issue by cmurf** [iovisor/bcc#3869]
Stack track from `biosnoop` `biolatency` and `biotop` are the same on both Fedora 35 and 36 (pre-release), with bcc-tools-0.21.0-1.fc35.x86_64 and bcc-tools-0.22.0-3.fc36.x86_64. Nothing found in dmesg.

```
$ sudo /usr/share/bcc/tools/biosnoop
/virtual/main.c:54:24: error: incomplete definition of type 'struct request'
        .data_len = req->__data_len
                    ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
/virtual/main.c:89:26: error: incomplete definition of type 'struct request'
        data.sector = req->__sector;
                      ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
/virtual/main.c:91:38: error: incomplete definition of type 'struct request'
        struct gendisk *rq_disk = req->rq_disk;
                                  ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
/virtual/main.c:108:26: error: incomplete definition of type 'struct request'
    data.rwflag = !!((req->cmd_flags & REQ_OP_MASK) == REQ_OP_WRITE);
                      ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
4 errors generated.
Traceback (most recent call last):
  File "/usr/share/bcc/tools/biosnoop", line 165, in <module>
    b = BPF(text=bpf_text)
  File "/usr/lib/python3.10/site-packages/bcc/__init__.py", line 452, in __init__
    raise Exception("Failed to compile BPF module %s" % (src_file or "<text>"))
Exception: Failed to compile BPF module <text>
$ sudo /usr/share/bcc/tools/biolatency
cannot attach kprobe, probe entry may not exist
Traceback (most recent call last):
  File "/usr/share/bcc/tools/biolatency", line 176, in <module>
    b.attach_kprobe(event="blk_account_io_done",
  File "/usr/lib/python3.10/site-packages/bcc/__init__.py", line 803, in attach_kprobe
    raise Exception("Failed to attach BPF program %s to kprobe %s" %
Exception: Failed to attach BPF program b'trace_req_done' to kprobe b'blk_account_io_done'
$ sudo /usr/share/bcc/tools/biotop
/virtual/main.c:55:24: error: incomplete definition of type 'struct request'
        .data_len = req->__data_len
                    ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
/virtual/main.c:78:21: error: incomplete definition of type 'struct request'
    info.major = req->rq_disk->major;
                 ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
/virtual/main.c:79:21: error: incomplete definition of type 'struct request'
    info.minor = req->rq_disk->first_minor;
                 ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
/virtual/main.c:92:26: error: incomplete definition of type 'struct request'
    info.rwflag = !!((req->cmd_flags & REQ_OP_MASK) == REQ_OP_WRITE);
                      ~~~^
include/linux/blkdev.h:24:8: note: forward declaration of 'struct request'
struct request;
       ^
4 errors generated.
Traceback (most recent call last):
  File "/usr/share/bcc/tools/biotop", line 177, in <module>
    b = BPF(text=bpf_text)
  File "/usr/lib/python3.10/site-packages/bcc/__init__.py", line 452, in __init__
    raise Exception("Failed to compile BPF module %s" % (src_file or "<text>"))
Exception: Failed to compile BPF module <text>
```

btrfsslower works ok so not all bcc-tools are affected
```
$ sudo /usr/share/bcc/tools/btrfsslower
Tracing btrfs operations slower than 10 ms
TIME     COMM           PID    T BYTES   OFF_KB   LAT(ms) FILENAME
^C$ 
```

kernel-5.16.10-200.fc35.x86_64
kernel-core-5.16.10-200.fc35.x86_64
kernel-devel-5.16.10-200.fc35.x86_64
kernel-modules-5.16.10-200.fc35.x86_64
bcc-tools-0.21.0-1.fc35.x86_64
bison-3.7.6-3.fc35.x86_64
clang12-libs-12.0.1-3.fc35.x86_64
flex-2.6.4-9.fc35.x86_64
llvm12-libs-12.0.1-2.fc35.x86_64
m4-1.4.19-2.fc35.x86_64
elfutils-libelf-devel-0.186-1.fc35.x86_64

kernel-debug-5.17.0-0.rc5.102.fc37.x86_64
kernel-debug-core-5.17.0-0.rc5.102.fc37.x86_64
kernel-debug-devel-5.17.0-0.rc5.102.fc37.x86_64
kernel-debug-modules-5.17.0-0.rc5.102.fc37.x86_64
bcc-tools-0.22.0-3.fc36.x86_64
bison-3.8.2-2.fc36.x86_64
clang-libs-13.0.1-1.fc36.x86_64
flex-2.6.4-10.fc36.x86_64
llvm-libs-13.0.1-1.fc36.x86_64
m4-1.4.19-3.fc36.x86_64
elfutils-libelf-devel-0.186-1.fc36.x86_64

On Fedora 36 (pre-release)
```
$ journalctl -k | grep -i 'bpf\|bcc'
Feb 21 10:44:36 fedora kernel: LSM support for eBPF active
Feb 21 10:44:37 fedora systemd[1]: systemd v250.3-4.fc36 running in system mode (+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
Feb 21 10:44:37 fedora systemd[1]: Got EBADF when using BPF_F_ALLOW_MULTI, which indicates it is supported. Yay!
Feb 21 10:44:37 fedora systemd[1]: Controller 'bpf-firewall' supported: yes
Feb 21 10:44:37 fedora systemd[1]: Controller 'bpf-devices' supported: yes
Feb 21 10:44:37 fedora systemd[1]: Controller 'bpf-foreign' supported: yes
Feb 21 10:44:37 fedora systemd[1]: Controller 'bpf-socket-bind' supported: yes
Feb 21 10:44:37 fedora systemd[1]: Controller 'bpf-restrict-network-interfaces' supported: yes
Feb 21 10:44:37 fedora systemd[1]: LSM BPF program attached
Feb 21 10:44:41 fedora systemd[1]: systemd v250.3-4.fc36 running in system mode (+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified)
Feb 21 10:44:41 fedora systemd[1]: Got EBADF when using BPF_F_ALLOW_MULTI, which indicates it is supported. Yay!
Feb 21 10:44:41 fedora systemd[1]: Controller 'bpf-firewall' supported: yes
Feb 21 10:44:41 fedora systemd[1]: Controller 'bpf-devices' supported: yes
Feb 21 10:44:41 fedora systemd[1]: Controller 'bpf-foreign' supported: yes
Feb 21 10:44:41 fedora systemd[1]: Controller 'bpf-socket-bind' supported: yes
Feb 21 10:44:41 fedora systemd[1]: Controller 'bpf-restrict-network-interfaces' supported: yes
Feb 21 10:44:41 fedora systemd[1]: LSM BPF program attached
```

**Comment by kreddkrikk** [comment]
It looks like `struct request` was moved from `linux/blkdev.h` to `linux/blk-mq.h` in kernel 5.16. When I replace `#include <linux/blkdev.h>` with `#include <linux/blk-mq.h>` in `bpf_test` in the `biosnoop` script, it fixes the problem for me.

**Comment by yonghong-song** [comment]
@cmurf @kreddkrikk could you help with a pull request to fix the issue? In this case, I think using LINUX_VERSION_CODE should be okay.

**Comment by kreddkrikk** [comment]
@yonghong-song Adding a conditional directive for the header includes also fixes the problem for me (using kernel 5.16.10):

```
#include <uapi/linux/ptrace.h>
#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 16, 0)
#include <linux/blkdev.h>
#else
#include <linux/blk-mq.h>
#endif
```

**Comment by jeromemarchand** [comment]
blk-mq.h header has always included blkdev.h anyway. So just replacing blkdev by blk-mq should work just as well on older and newer version of the kernel.

**Comment by jeromemarchand** [comment]
Also it looks like biolatpcts used to suffer from the issue too, but it was fixed by commit 11614bcacdecd4, in this case by including blk-mq.h besides blkdev.h.
